### ISSUE TITLE
The filtered positional operator $[<identifier>] (arrayFilters) does not work in Mongo update.

### ISSUE THREAD

**Issue by gustawdaniel** [meteor/meteor#10103]
❗️ Bug or missing docs about array filter option in Mongo update.

Meteor version: `Meteor 1.7.0.3`

## Short description

In mongo 3.6 there is avaliable filtered positional operator `$[<identifier>]`. Described here:

> https://docs.mongodb.com/manual/reference/operator/update/positional-filtered/

Using this was not possible before `Meteor v1.7` because of `mongodb` driver had version `2`. From `Meteor v1.7` we updated `mongodb` to `3` so implementation of `positional filter operator` should be possible now.

This problem was mentioned in issue:

> https://github.com/meteor/meteor-feature-requests/issues/268

Although of update `mongodb` from `2.2.34` to `3.0.5` in https://github.com/meteor/meteor/pull/9790/files `filtered positional operator` still not works.

## How it should work (mongo shell)

Login to mongo shell and create simplest possible resource that allow to reproduce problem eg.:

    db.insert({_id: 'crons'}, {_id: 'crons', value: [{name: 'getList', interval: 1000},{name: 'getOffer', interval: 1000}]});

Next use `filtered positional operator $[<identifier>]` to observe how it should work:

    db.configs.update({_id:'crons'},{ $set: { "value.$[elem].interval": 0 } },{ arrayFilters: [ {"elem.name": "getOffer"} ] })

Now you should see that resource looks like this:

    {
	    "_id" : "crons",
	    "value" : [ { "name" : "getList", "interval" : 1000 },
		        { "name" : "getOffer", "interval" : 0 } ]
    }

You can see that interval was changed for value with name "getOffer".

## How it **does not** work with meteor

Create simple method that repeats updated with arrayFilters key in third argument of update:

```
Meteor.methods({
        'config.update-cron-interval'({name, interval}) {
            check(name, String);
            check(interval, Number);

            Configs.update(
                {_id: 'crons'},
                { $set: { "value.$[elem].interval": interval } },
                {arrayFilters: [{"elem.name": name}]}
            );
        }
    })
```

Calling this method throws error:

```
I20180727-16:51:49.895(2)? { MongoError: No array filter found for identifier 'elem' in path 'value.$[elem].interval'
I20180727-16:51:49.895(2)?     at Function.MongoError.create (/home/daniel/.meteor/packages/npm-mongo/.3.0.7.bpkgyk.3f1fv++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/mongodb-core/lib/error.js:45:10)
I20180727-16:51:49.896(2)?     at toError (/home/daniel/.meteor/packages/npm-mongo/.3.0.7.bpkgyk.3f1fv++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/mongodb/lib/utils.js:149:22)
I20180727-16:51:49.896(2)?     at /home/daniel/.meteor/packages/npm-mongo/.3.0.7.bpkgyk.3f1fv++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/mongodb/lib/collection.js:1029:39
I20180727-16:51:49.896(2)?     at /home/daniel/.meteor/packages/npm-mongo/.3.0.7.bpkgyk.3f1fv++os+web.browser+web.browser.legacy+web.cordova/npm/node_modules/mongodb-core/lib/connection/pool.js:544:18
I20180727-16:51:49.896(2)?     at _combinedTickCallback (internal/process/next_tick.js:131:7)
I20180727-16:51:49.897(2)?     at process._tickDomainCallback (internal/process/next_tick.js:218:9)
I20180727-16:51:49.897(2)?   name: 'MongoError',
I20180727-16:51:49.897(2)?   message: 'No array filter found for identifier \'elem\' in path \'value.$[elem].interval\'',
I20180727-16:51:49.897(2)?   driver: true,
I20180727-16:51:49.897(2)?   index: 0,
I20180727-16:51:49.898(2)?   code: 2,
I20180727-16:51:49.898(2)?   errmsg: 'No array filter found for identifier \'elem\' in path \'value.$[elem].interval\'' }
```

## Temporary workaround

Instead mentioned update we can do the following things

1) query `findOne`
2) transform object in javascript
3) query `update`

In this case we can do:

    const crons = Configs.findOne({_id: 'crons'});

    crons.value = crons.value.map(e => { if(e.name===name) { e.interval = interval } return e; });

    Configs.update({_id: 'crons'}, crons );

But it would be better if this feature will be implemented or lacking implementation could be mentioned in docs.

**Comment by Nader182** [comment]
Have you tried accessing the raw collection like so

` Configs.rawCollection().update(
                {_id: 'crons'},
                { $set: { "value.$[elem].interval": interval } },
                {arrayFilters: [{"elem.name": name}]}
            );`

I've managed to get this working in new Meteor 1.7 projects, however projects that have been UPGRADED do not use arrayFilters at all :/

**Comment by inoyatov** [comment]
@Nader182 it works like charm. I spend two days in order to find solution. Finally I realized that is not supported and only workaround is using rawCollection. :) I hope they fill fix it soon.

**Comment by mitar** [comment]
There is no third argument to `update` in Meteor. This is why `{arrayFilters: [{"elem.name": name}]}` is simply not passed to underlying `update` call, so MongoDB does not find it.

This should be relatively easy to fix/add.

**Comment by mitar** [comment]
See also #9721.

**Comment by afrokick** [comment]
Any news?

in 1.8 it still doesn't work.

**Comment by zhukovsp** [comment]
@benjamn Hello! Do you have any news about this feature?

**Comment by mitar** [comment]
Still relevant.

**Comment by wallind** [comment]
Hi all! My company uses Meteor/Mongo as our backend stack and I just ran into an use case for this.
@mitar do you know if this is still being worked on? Otherwise maybe I'll take a stab at it 😏

P.S. I love your avatar

**Comment by filipenevola** [comment]
hi @wallind feel free to work on this.

**Comment by filipenevola** [comment]
Published mongo@1.9.1.

**Comment by Slavrix** [comment]
@filipenevola champion.

does anyone have any experience with updating the @types?
